### PR TITLE
Add Process support for UEFI

### PR DESCRIPTION
**Pull request by Ayush1325**
UEFI does not have an actual process. However, it does provide methods to launch and execute another UEFI image. Having process support is important since it is possible to run rust test suit using `Command::output` and is the first step towards being able to run it for UEFI.

Here is an overview of how the support is implemented.

- We create a copy of the SystemTable. This is required since at least OVMF seems to crash if the original system table is modified.
- Stdout and Stderr pipe works by registering a new `simple_text_output` Protocol and pointing the child system table to use those.
- `Stdio::Inherit` just points the console to the current running image console which seems to work with even 3 levels of process.
- `spawn` is left unimplemented since it does not make sense for UEFI architecture. Additionally, since https://github.com/rust-lang/rust/pull/105458 was merged, the `spawn` and `output` implementations are completely independent.

### CODE REVIEW

**Comment by Nilstrieb** [comment]
@dvdhrm @nicholasbishop could I get a review from you?

File: library/std/src/sys/pal/uefi/process.rs
Comment (Nilstrieb) [inline library/std/src/sys/pal/uefi/process.rs:38]: What's stopping you from doing this in this PR?

Comment (Ayush1325) [reply library/std/src/sys/pal/uefi/process.rs:38]: I will remove that comment. It's a remnant from `unsupported/process.rs`. It is actually possible to use `Stdio::Inherit`, `Stdio::Null`, `Stdio::MakePipe` in UEFI.

File: library/std/src/sys/pal/uefi/process.rs
Comment (joboet) [inline library/std/src/sys/pal/uefi/process.rs:421]: I think these updates need to adjust the CRC in the table header, otherwise very paranoid images will complain.

File: library/std/src/sys/pal/uefi/process.rs
Comment (joboet) [inline library/std/src/sys/pal/uefi/process.rs:386]: To preserve forward-compatibility if new fields are added to the system table, I think we should copy `system_table.hdr.header_size` bytes instead of the fields defined  in `r_efi`.

File: library/std/src/sys/pal/uefi/process.rs
Comment (joboet) [inline library/std/src/sys/pal/uefi/process.rs:551]: This isn't sound because `Box` is `noalias`, so this reference will be invalidated once `mode` is moved
```suggestion
                mode: Box::into_raw(mode),
```
and add a `Drop` implementation that calls `Box::from_raw` to deallocate.

File: library/std/src/sys/pal/uefi/process.rs
Comment (joboet) [inline library/std/src/sys/pal/uefi/process.rs:576]: Likewise.

File: library/std/src/sys/pal/uefi/process.rs
Comment (joboet) [inline library/std/src/sys/pal/uefi/process.rs:389]: As described below, this isn't sound, as this reference will be invalidated as soon as `st` is returned.

File: library/std/src/sys/pal/uefi/process.rs
Comment (joboet) [inline library/std/src/sys/pal/uefi/process.rs:416]: This doesn't do anything, `usize` and `*mut u16` don't implement `Drop`. We should call the allocator instead.

**Review by joboet — CHANGES_REQUESTED** [review]
There are a number of cases where we open a protocol, but never close it. Maybe a `BorrowedProtocol` wrapper with the necessary `Drop` implementation could help solve this?

**Comment by Nilstrieb** [comment]
r? joboet

**Comment by Ayush1325** [comment]
> There are a number of cases where we open a protocol, but never close it. Maybe a BorrowedProtocol wrapper with the necessary Drop implementation could help solve this?

We are using `GET_PROTOCOL` attribute to open protocol, which does not require us to close the protocol. Feel free to point out places that might benefit from open protocol in a different way since those will require BorrowedProtocol wrapper.

File: library/std/src/sys/pal/uefi/process.rs
Comment (Ayush1325) [reply library/std/src/sys/pal/uefi/process.rs:416]: Wouldn't MaybeUninit<usize> be allocated on stack? Only exit data is allocated by the other image we launch.

File: library/std/src/sys/pal/uefi/process.rs
Comment (Ayush1325) [reply library/std/src/sys/pal/uefi/process.rs]: How do you propose going about this? For the compiler, `T` is `Sized`, but we want to construct and use it as if `T` is `!Sized`. So I am not quite sure how to proceed. Maybe create an `OwnedTable`? It doesn't seem like I can use standard Box for this.

**Review by joboet — COMMENTED** [review]
I think this PR could be simplified quite a bit, which would also help with soundness:

Currently, `Command` in `uefi_command_internal` stores things like the system table, stdout, stderr, but it also references these resources in the image load options, or the system table. So in effect, we have a self-referential structure. Those are really tricky to get right, and this PR has some soundness bugs related to them.

Luckily, we can morph this into a Rust-like form, because all these things only need to live for the duration of the `output` call. So I'd recommend just storing these resources in variables inside `output`, and then referencing them at the appropriate places:
```rust
fn output(&mut self) {
    let stdout = create_pipe(...);
    let stderr = create_pipe(...);
    let args = convert_args(...);
    let mut st = create_st(&stdout, &stderr);
    run(&args, &mut st);
}
```

@rustbot author

**Comment by Dylan-DPC** [comment]
@Ayush1325 any updates on this? thanks

**Comment by Ayush1325** [comment]
> @Ayush1325 any updates on this? thanks

Hi. Sorry for the delay. My college exams were going on and am busy with some contract work. Will try to address everything in a few days

File: library/std/src/sys/pal/uefi/process.rs
Comment (Ayush1325) [reply library/std/src/sys/pal/uefi/process.rs:305]: This does not seem to work:
```shell
error[E0308]: mismatched types
   --> library/std/src/sys/pal/uefi/process.rs:305:9
    |
304 |     fn next(&mut self) -> Option<&'a OsStr> {
    |                           ----------------- expected `core::option::Option<&'a OsStr>` because of return type
305 |         self.iter.next().as_deref()
    |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `Option<&OsStr>`, found `Option<&OsString>`
    |
    = note: expected enum `core::option::Option<&'a OsStr>`
               found enum `core::option::Option<&OsString>`

For more information about this err
```

Comment (joboet) [reply library/std/src/sys/pal/uefi/process.rs:294]: Ah right. Nevermind.

File: library/std/src/sys/pal/uefi/helpers.rs
Comment (Ayush1325) [reply library/std/src/sys/pal/uefi/helpers.rs:337]: Will switch to `uninstall_multiple_protocol_interfaces` once https://github.com/r-efi/r-efi/pull/79 is merged.

File: library/std/src/sys/pal/uefi/process.rs
Comment (Ayush1325) [reply library/std/src/sys/pal/uefi/process.rs]: I have added an `OwnedTable` type

File: library/std/src/sys/pal/uefi/helpers.rs
Comment (dvdhrm) [reply library/std/src/sys/pal/uefi/helpers.rs:337]: I dont think varargs are supported for eficall. I think it is only supported for "extern C". So the multi-call is not an option.

Comment (Ayush1325) [reply library/std/src/sys/pal/uefi/helpers.rs:337]: Yes, but it seems `uninstall_multiple_protocol_interfaces` (and `install_multiple_protocol_interfaces`) are to be preferred over their single counterparts due to EFI 1.1 extensions that allow tracking. Thus `uninstall_multiple_protocol_interfaces` only fails if `One of the protocol interfaces was not previously installed on Handle. `, at least according to spec.

Comment (dvdhrm) [reply library/std/src/sys/pal/uefi/helpers.rs:337]: What I am saying is that `uninstall_multiple_protocol_interfaces` uses a variable number of arguments, which is not supported by stable Rust for `eficall`. The prototype provided by `r-efi` is not usable in its current form and can easily lead to ABI violations.

You can look at EDK2 and see how `uninstall_multiple_protocol_interfaces` is implemented, and simply copy the logic from there.

Comment (dvdhrm) [reply library/std/src/sys/pal/uefi/helpers.rs:337]: Also see: https://github.com/rust-lang/rust/issues/100189

Since `std` can use unstable features, you can just provide your own `fn` type and transmute the fn-pointer from the boot-services table.

**Review by joboet — COMMENTED** [review]
Sorry for the delay, haven't had that much time recently. It's looking quite good and sound, but some smaller problems remain.

File: library/std/src/sys/pal/uefi/process.rs
Comment (Ayush1325) [reply library/std/src/sys/pal/uefi/process.rs:248]: Yes, it should be done, but this is probably not the PR to do it.

File: library/std/src/sys/pal/uefi/process.rs
Comment (Ayush1325) [reply library/std/src/sys/pal/uefi/process.rs]: Do you mean I should extract argument parsing to it's own function and leave the current logic, or just have a blank implementation for handling arguments for now?

Also, what are the rules regarding args?
1. Is the user supposed to put `"` around string arguments (those that contain whitespace). Or is the implementation supposed to detect if quote is needed?
2. Is the user supposed to escape the characters using `^` (UEFI escape), or should the implementation detect characters that need to be escaped?
3. What happens if user passes a NULL character?

File: library/std/src/sys/pal/uefi/helpers.rs
Comment (joboet) [inline library/std/src/sys/pal/uefi/helpers.rs:288]: The protocol is a variable-width structure, but this returns a reference to a fixed-width node. IIRC, it's still unclear whether taking the reference restricts the provenance of the reference to the reference type, but until that is figured out, I'd prefer if this were expressed as an `as_ptr()` or `as_mut_ptr()` method instead.

File: library/std/src/sys/pal/uefi/helpers.rs
Comment (joboet) [inline library/std/src/sys/pal/uefi/helpers.rs:329]: We need to drop the allocated protocol here.
```suggestion
            drop(Box::from_raw(protocol));
            return Err(crate::io::Error::from_raw_os_error(r.as_usize()));
```

File: library/std/src/sys/pal/uefi/process.rs
Comment (joboet) [inline library/std/src/sys/pal/uefi/process.rs:52]: Could you, for the moment, stub this out? This will not behave as expected.
```suggestion
        panic!("unsupported")
```

File: library/std/src/sys/pal/uefi/process.rs
Comment (joboet) [inline library/std/src/sys/pal/uefi/process.rs:361]:
```suggestion
                let st =
```

File: library/std/src/sys/pal/uefi/process.rs
Comment (joboet) [reply library/std/src/sys/pal/uefi/process.rs]: We need to do the escaping so that `env::args()` in the image can recover the exact arguments passed in here. As for NULL, UNIX remembers whether one of the arguments contains NULL inside the `Command` structure and returns an error upon spawning in that case.

File: library/std/src/sys/pal/uefi/helpers.rs
Comment (joboet) [inline library/std/src/sys/pal/uefi/helpers.rs:239]: This will ignore internal NULL bytes in the string, resulting in truncation.

**Review by joboet — COMMENTED** [review]
Just one bug left :tada:. I'll be filing an ACP for introducing a `RawExitStatus` and wait for it to be discussed on Tuesday before merging this, just so we are not locking in any choices, but otherwise this seems good to go.

File: library/std/src/sys/pal/uefi/helpers.rs
Comment (Ayush1325) [reply library/std/src/sys/pal/uefi/helpers.rs:239]: I am not sure I understand. Can you provide me with an example?

Comment (joboet) [reply library/std/src/sys/pal/uefi/helpers.rs:239]: Exactly, yes. We should return an error in that case instead, [like on Windows](https://github.com/rust-lang/rust/blob/5ced3dad5752cc9423a19e9a5aafad6e12ed3063/library/std/src/sys/pal/windows/mod.rs#L177-L196).

Comment (Ayush1325) [reply library/std/src/sys/pal/uefi/helpers.rs:239]: I am not doing manual loop unrolling like the windows version since I am not sure if it is necessary.

Comment (workingjubilee) [reply library/std/src/sys/pal/uefi/helpers.rs:239]: don't worry. the loop unrolling was from quite some time ago. we should test to see if it even does anything anymore.

**Review by joboet — APPROVED** [review]
I think this is ready to go. Let's wait until https://github.com/rust-lang/libs-team/issues/403 is resolved (there was no libs-api meeting yesterday) and then merge this.

**Comment by joboet** [comment]
Thank you and sorry for the delay!
@bors r+

**Comment by tgross35** [comment]
This failed the rollup because we are starting to forbid `unsafe_op_in_unsafe_fn`, https://github.com/rust-lang/rust/pull/127954#issuecomment-2238676554. It looks like this just misses a few cases, you'll probably start getting the warnings locally if you rebase.

@bors r-

**Comment by workingjubilee** [comment]
This looks basically like how I last saw it, yeah.

@bors r=joboet

**Comment by Ayush1325** [comment]
> This looks basically like how I last saw it, yeah.
> 
> @bors r=joboet

Yup, there was only 1 place of failure after rebase: https://github.com/rust-lang/rust/pull/123196/files#diff-21f19573f63f3900715f66b8ad29e271024c92279c5d73e0b851af423d213129R328

**Pull request merged**
